tianshou/data/batch.py

```diff
@@ -16,10 +16,10 @@
 
 def _is_batch_set(data: Any) -> bool:
     if isinstance(data, (list, tuple)):
-        if len(data) > 0 and isinstance(data[0], (dict, Batch)):
+        if len(data) > 0 and all(isinstance(e, (dict, Batch)) for e in data):
             return True
-    elif isinstance(data, np.ndarray):
-        if isinstance(data.item(0), (dict, Batch)):
+    elif isinstance(data, np.ndarray) and data.dtype == np.object:
+        if all(isinstance(e, (dict, Batch)) for e in data.tolist()):
```

Comment:
Same, `tolist` is doing a copy, which is time consuming.

I think you misunderstand the point of checking types: type checking is used to determine the appropriate behavior, NOT to check that the user is doing something crazy. You should be consider the user as an adversary. Checking everything is taking a lot of time, for little gain.

Reply:
I think you misunderstand the spent time. Deepcopy is time-consuming, but here we assure it is an array with ``np.object`` type, so ``tolist()`` only copies a reference.

There is indeed a trade-off between flexibility and debug-ability. But these operators are cheap in terms of time-cost, and can make sure users will not be confused by strange exceptions.

I'm not considering the user as an adversary, I'm avoiding users from unexpected behaviors with extremely long traceback. When they code something wrong, they should get the hint about what is going wrong.

Reply:
> I think you misunderstand the spent time. Deepcopy is time-consuming, but here we assure it is an array with np.object type, so tolist() only copies a reference.

True

> I'm not considering the user as an adversary, I'm avoiding users from unexpected behaviors with extremely long traceback. When they code something wrong, they should get the hint about what is going wrong.

It is straightforward from the documentation that it does not make sense to use something else than `np.ndarray` of `Batch`/`dict` without keyword. I don't think you should infantilize people at the point of not reading the first line of the documentation. Even though it is cheap, we are creating `Batch` millions of times during the whole training process, so at the end it matters.